### PR TITLE
 ✨ New `pay with payment-method` implementation

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

**Reimplements  "pay with payment-method"** since app-motion changed the design from *Init-Prompt-Ack* worfklow to a direct http request-response. SEE all changes of openapi-specs in https://github.com/ITISFoundation/appmotion-exchange/issues/12

- `webserver`:
   -  ✨ keeps the same interface for `POST /wallets/{w_id}/payments-method/{pm_id}:pay` but changes implementation      
        - ack notification done as a fire&forget
    - ✨ new `pay_with_payment_method` on  _rpc_
- `payments`:
    - 🎨 changes in `PaymentMethodGet` model. Needed to remove some info from the front-end
    - ✨ new `pay_with_payment_method` on  _rpc_ and _gateway_ interfaces
    - improves tests against external (specifically app-motion's gateway service)
    - 🐛 fixes and changes in openapi specs for gateway and ack APIs (upon agreement with app-motion)
           - `AckPayment.provider_payment_id` to keep stripe payment identifier
           - `AckPaymentWithPaymentMethod.payment_id` 
    - 🎨  _gateway_ error handling
- 🔨 useful extensions for openapi specs and makefile

## Related issue/s

- Part of #4657 

## How to test

Driving tests are
```
cd services/web/server
make install-dev
pytest -vv -k test_one_time_payment_with_payment_method tests/unit
```
and
```
cd services/payments
make install-dev
pytest -vv -k test_webserver_pay_with_payment_method_workflow tests/unit
```

## Dev Checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

## DevOps 
None

### CODE REVIEW

File: api/specs/web-server/_wallets.py [hunk 1]
```diff
@@ -168,10 +168,10 @@ async def delete_payment_method(
 @router.post(
     "/wallets/{wallet_id}/payments-methods/{payment_method_id}:pay",
     response_model=Envelope[WalletPaymentInitiated],
-    response_description="Payment initialized",
-    status_code=status.HTTP_202_ACCEPTED,
+    response_description="Pay with payment-method",
+    status_code=status.HTTP_202_ACCEPTED,  # TODO: check in front-end
```
Comment (sanderegg) [inline api/specs/web-server/_wallets.py:172]: forgotten todo?

File: packages/models-library/src/models_library/api_schemas_webserver/wallets.py [hunk 2]
```diff
@@ -124,14 +124,11 @@ class Config(OutputSchema.Config):
...
-    expiration_month: int
-    expiration_year: int
-    street_address: str
-    zipcode: str
-    country: str
+    card_holder_name: str | None = None
+    card_number_masked: str | None = None
+    card_type: str | None = None
+    expiration_month: int | None = None
+    expiration_year: int | None = None
```
Comment (sanderegg) [inline packages/models-library/src/models_library/api_schemas_webserver/wallets.py:131]: someone told me to be careful with None ;) kidding it's late...

File: services/payments/src/simcore_service_payments/api/rpc/_payments_methods.py [hunk 3]
```diff
@@ -127,6 +128,11 @@ async def init_payment_with_payment_method(  # noqa: PLR0913 # pylint: disable=t
     user_email: EmailStr,
     comment: str | None = None,
 ) -> WalletPaymentInitiated:
+    warnings.warn(
+        f"{__name__}.init_payment_with_payment_method is deprecated. Use instead pay_with_payment_method",
```
Comment (sanderegg) [inline services/payments/src/simcore_service_payments/api/rpc/_payments_methods.py:132]: is this method still needed? if not I would prefer to see it disappear. if it will disappear later, then maybe add in the warning when it is safe to do so

File: services/payments/src/simcore_service_payments/models/schemas/acknowledgements.py [hunk 4]
```diff
@@ -9,6 +10,17 @@ class _BaseAck(BaseModel):
...
 
 
+class _BaseAckPayment(_BaseAck):
+    provider_payment_id: IDStr = Field(
+        default=None,
+        description="Payment ID from the provider (e.g. stripe payment ID)",
+    )
+
+    invoice_url: HttpUrl | None = Field(
+        default=None, description="Link to invoice is required when success=true"
```
Comment (sanderegg) [inline services/payments/src/simcore_service_payments/models/schemas/acknowledgements.py:20]: validator?

File: services/payments/src/simcore_service_payments/services/payments.py [hunk 5]
```diff
@@ -189,6 +197,12 @@ async def init_payment_with_payment_method(
     user_email: EmailStr,
     comment: str | None = None,
 ) -> WalletPaymentInitiated:
+    warnings.warn(
+        f"{__name__}.init_payment_with_payment_method is deprecated. Use instead pay_with_payment_method",
+        DeprecationWarning,
+        stacklevel=1,
+    )
```
Comment (sanderegg) [inline services/payments/src/simcore_service_payments/services/payments.py:204]: same remarks here.

File: services/payments/tests/unit/conftest.py [hunk 6]
```diff
@@ -126,15 +129,22 @@ def payments_clean_db(postgres_db: sa.engine.Engine) -> Iterator[None]:
...
     async with LifespanManager(
-        test_app,
-        startup_timeout=None,  # for debugging
-        shutdown_timeout=10,
+        the_test_app,
+        startup_timeout=None if is_pdb_enabled else MAX_TIME_FOR_APP_TO_STARTUP,
+        shutdown_timeout=None if is_pdb_enabled else MAX_TIME_FOR_APP_TO_SHUTDOWN,
     ):
-        yield test_app
+        yield the_test_app
```
Comment (sanderegg) [inline services/payments/tests/unit/conftest.py:147]: pretty. this could probably go in pytest-simcore at some point...

File: services/payments/Makefile [hunk 7]
```diff
@@ -23,8 +23,8 @@ external ?= .env-secret
 
 test-dev-unit-external: ## runs test-dev against external service defined in $(external) envfile
 	# Running tests using external environ '$(external)'
-	$(MAKE) test-dev-unit pytest-parameters="--external-envfile=$(external)"
+	$(MAKE) test-dev-unit pytest-parameters="--external-envfile=$(external) -m can_run_against_external"
```
Comment (sanderegg) [inline services/payments/Makefile:26]: nice

File: services/web/server/src/simcore_service_webserver/payments/_onetime_api.py [hunk 8]
```diff
@@ -210,6 +210,7 @@ async def _ack_creation_of_wallet_payment(
     completion_state: PaymentTransactionState,
     message: str | None = None,
     invoice_url: HttpUrl | None = None,
+    nofity_enabled: bool = True,
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/payments/_onetime_api.py:213]:
```suggestion
    notify_enabled: bool = True,
```

File: services/web/server/src/simcore_service_webserver/payments/_onetime_api.py [hunk 9]
```diff
@@ -285,6 +289,100 @@ async def cancel_payment_to_wallet(
...
+        app,
+        payment_id=inited.payment_id,
+        completion_state=PaymentTransactionState.SUCCESS,
+        message=f"Fake payment completed with payment-id = {payment_method_id}",
+        invoice_url=f"https://fake-invoice.com/?id={inited.payment_id}",  # type: ignore
+        nofity_enabled=False,
+    )
+
+
+async def pay_with_payment_method(
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/payments/_onetime_api.py:329]: we are lucky we do not pay with paper ;)

**Review by sanderegg — APPROVED** [review]
looking good!

File: services/payments/src/simcore_service_payments/models/schemas/acknowledgements.py (hunk 4 above)
Comment (pcrespov) [reply services/payments/src/simcore_service_payments/models/schemas/acknowledgements.py:20]: there is already one in place

File: packages/pytest-simcore/src/pytest_simcore/helpers/rawdata_fakers.py [hunk 10]
```diff
@@ -251,16 +251,14 @@ def random_api_key(product_name: str, user_id: int, **overrides) -> dict[str, An
 def random_payment_method_data(**overrides) -> dict[str, Any]:
     # Produces data for GetPaymentMethod
     data = {
-        "idr": FAKE.uuid4(),
+        "id": FAKE.uuid4(),
```
Comment (matusdrobuliak66) [inline packages/pytest-simcore/src/pytest_simcore/helpers/rawdata_fakers.py:254]: why this change? isn't `idr` correct?

File: services/payments/src/simcore_service_payments/services/payments.py [hunk 11]
```diff
@@ -206,23 +213,41 @@ async def init_payment_with_payment_method(
...
+    payment_id = ack.payment_id
 
-    return WalletPaymentInitiated(
-        payment_id=f"{payment_id}",
-        payment_form_url=None,
+    async for attempt in AsyncRetrying(
+        stop=stop_after_attempt(3),
+        retry=retry_if_exception_type(PaymentAlreadyExistsError),
+        reraise=True,
+    ):
```
Comment (matusdrobuliak66) [inline services/payments/src/simcore_service_payments/services/payments.py:222]: Q: Why this retry is here? When can this happen? Thanks

File: packages/pytest-simcore/src/pytest_simcore/helpers/rawdata_fakers.py (hunk 10 above)
Comment (matusdrobuliak66) [reply packages/pytest-simcore/src/pytest_simcore/helpers/rawdata_fakers.py:254]: Aha I see it now :+1:

**Review by matusdrobuliak66 — APPROVED** [review]
Thanks! I will adjust my PR to this one. I have just one question regarding the retry statement.

File: services/payments/src/simcore_service_payments/models/schemas/acknowledgements.py [hunk 12]
```diff
@@ -85,6 +94,24 @@ def success_requires_invoice(cls, v, values):
         return v
 
 
+class AckPaymentWithPaymentMethod(_BaseAckPayment):
+    # NOTE: This model is equivalent to `AckPayment`, nonetheless
+    # I decided to separate it for clarity in the OAS since in payments
+    # w/ payment-method the field `saved` will never be provided,
```
Comment (GitHK) [inline services/payments/src/simcore_service_payments/models/schemas/acknowledgements.py:100]:
```suggestion
    # NOTE: This model is equivalent to `AckPayment`. Separated for clarity in the OAS.
    # In "payments without payment-method" the field `saved` will never be provided.
```

File: services/payments/src/simcore_service_payments/services/payments.py [hunk 13]
```diff
@@ -206,23 +213,41 @@ async def init_payment_with_payment_method(
...
-        comment=comment,
-        initiated_at=initiated_at,
-    )
+    payment_id = ack.payment_id
 
-    return WalletPaymentInitiated(
-        payment_id=f"{payment_id}",
-        payment_form_url=None,
+    async for attempt in AsyncRetrying(
+        stop=stop_after_attempt(3),
```
Comment (GitHK) [inline services/payments/src/simcore_service_payments/services/payments.py:219]: Suggestion: Since you are dealing with an external system that can be offline due to load, I think `wait_random_exponential` would also help here.

**Review by GitHK — APPROVED** [review]

File: services/payments/src/simcore_service_payments/services/payments.py (hunk 11 above)
Comment (pcrespov) [reply services/payments/src/simcore_service_payments/services/payments.py:222]: @matusdrobuliak66 Note that I create the uuid. If the uuid already exists, then it retries

File: services/payments/src/simcore_service_payments/services/payments.py (hunk 13 above)
Comment (pcrespov) [reply services/payments/src/simcore_service_payments/services/payments.py:219]: but this is just retrying  to ensure it has different uuids.
 if the system gets offline, i do not want it to retry
